添加：使用正则匹配 direct 和 proxy 列表中的host

halftan opened the pull request.

renzhn commented:
这样运行效率会大大降低吧

halftan commented:
添加了 Benchmark ，13寸 i5 rMBP 跑下来结果如下：

```
PASS
BenchmarkRegexp-4	2000000000	         0.00 ns/op
BenchmarkGoMap-4 	2000000000	         0.01 ns/op
ok  	github.com/renzhn/MEOW	1.184s
```

renzhn commented:
这里应该这么写吧：
```
func BenchmarkRegexp(b *testing.B) {
        list := newDomainList()
        list.initDomainList("doc/sample-config/direct_test", domainTypeDirect)
        list.initDomainList("doc/sample-config/proxy_test", domainTypeProxy)

        for i := 0; i < b.N; i++ {
                for url, _ := range testUrl {
                        list.regexJudge(url)
                }
        }
}

func BenchmarkOld(b *testing.B) {
        list := newDomainList()
        list.initDomainList("doc/sample-config/direct_test", domainTypeDirect)
        list.initDomainList("doc/sample-config/proxy_test", domainTypeProxy)

        for i := 0; i < b.N; i++ {
                for url, _ := range testUrl {
                        list.oldJudge(url)
                }
        }
}
```
我的测试结果如下：
```
PASS
BenchmarkRegexp	 1000000	      2122 ns/op
BenchmarkOld   	 2000000	       852 ns/op
ok  	MEOW	4.865s
```
而且正则表达式如果变复杂的话可能性能也会受到影响

renzhn commented:
而且旧的方法时间复杂度是不会随domain list增长而增长的

Kisesy commented:
加个选项吧，性能什么的倒是无所谓

halftan commented:
抱歉，对 golang 还不怎么熟悉 :sob:

不过我认为更灵活的过滤配置还是需要的，原来的方案并不能 cover 到所有的需求。

halftan commented:
可以为正则加一个特殊的配置语法，比如：
```
/s3.*amazonaws.*/
```
这样就可以减少正则匹配的次数，从而提高性能

renzhn commented:
MEOW 主要是通过 IP 来判断是否直连，例如 amazon s3 会被解析到国外 IP 从而使用代理连接，就我个人来说这套匹配机制目前还没有很大的问题。改成正则表达式后域名列表文件也要改，你可以再考虑考虑，欢迎交流~

renzhn commented:
@Kisesy 性能对我来说有所谓的，细小的差异累积起来会变成巨大的差异

halftan commented:
确实无法避免性能下降的问题。我现在使用正则的场景是 `apple.com` 直连，但是 `itunes.apple.com` 走代理，原先的配置方案无法满足这样的需求。不使用正则的话，是否可以将 Domain 的范围放宽到三级域名？

halftan commented:
d8aed69ac7ad942adbfbd44eaaf1f4fa9708ce61 中增加了正则表达式特殊配置语法，其他的配置依旧使用 hashmap 进行匹配，benchmark：
```
PASS
BenchmarkRegexp-4	  100000	     12385 ns/op
BenchmarkGoMap-4 	 1000000	      1339 ns/op
ok  	github.com/renzhn/MEOW	3.411s
```

PS: `testdata` 文件夹中添加了测试用 Domain list

renzhn commented:
benchmark的结果不忍直视啊，性能将近差了10倍！
https://golang.org/pkg/testing/ `ns/op`是指完成一次循环需要的时间

halftan commented:
我觉得不能从相差多少倍来看，毕竟增加了代码量，肯定会带来数十到数百ns的性能损耗。从时间来看，平均每次代理选择消耗的时间依旧在2微秒以内，我觉得可以接受

renzhn commented:
我在想为什么只有极少数的正则表达式规则却带来了这么大的性能差异，还需要完善和测试啊

halftan commented:
benchmark 用 URL 设置为单个，跑下来的结果是：
```
BenchmarkRegexp-4	  500000	      2325 ns/op
BenchmarkGoMap-4 	20000000	       104 ns/op
```

GoMap 的 Profile:
```
1900ms of 2080ms total (91.35%)
Dropped 13 nodes (cum <= 10.40ms)
Showing top 10 nodes out of 46 (cum >= 40ms)
      flat  flat%   sum%        cum   cum%
     630ms 30.29% 30.29%      780ms 37.50%  runtime.mapaccess1_faststr
     510ms 24.52% 54.81%      510ms 24.52%  runtime.mach_semaphore_signal
     230ms 11.06% 65.87%      630ms 30.29%  runtime.mallocgc
     130ms  6.25% 72.12%      130ms  6.25%  runtime.aeshashbody
     120ms  5.77% 77.88%     1630ms 78.37%  github.com/renzhn/MEOW.(*DomainList).oldJudge
      80ms  3.85% 81.73%       80ms  3.85%  runtime.heapBitsSetType
      80ms  3.85% 85.58%       80ms  3.85%  runtime.mach_semaphore_wait
      40ms  1.92% 87.50%     1670ms 80.29%  github.com/renzhn/MEOW.BenchmarkGoMap
      40ms  1.92% 89.42%      670ms 32.21%  runtime.newobject
      40ms  1.92% 91.35%       40ms  1.92%  runtime.prefetchnta
```

Regexp 的 Profile:
```
2240ms of 2600ms total (86.15%)
Dropped 17 nodes (cum <= 13ms)
Showing top 10 nodes out of 53 (cum >= 50ms)
      flat  flat%   sum%        cum   cum%
     920ms 35.38% 35.38%     1280ms 49.23%  regexp.(*machine).tryBacktrack
     310ms 11.92% 47.31%      400ms 15.38%  runtime.mCentral_Grow
     220ms  8.46% 55.77%      220ms  8.46%  regexp.(*bitState).push
     160ms  6.15% 61.92%      160ms  6.15%  runtime.mach_semaphore_signal
     140ms  5.38% 67.31%      140ms  5.38%  regexp.(*inputString).step
     130ms  5.00% 72.31%      130ms  5.00%  runtime.memclr
     130ms  5.00% 77.31%      130ms  5.00%  runtime.usleep
      90ms  3.46% 80.77%     1490ms 57.31%  regexp.(*machine).backtrack
      90ms  3.46% 84.23%       90ms  3.46%  sync/atomic.AddUint32
      50ms  1.92% 86.15%       50ms  1.92%  sync/atomic.CompareAndSwapUint32
```

从 Profile 的结果来看，`regexp.(*machine).tryBacktrack` 花费了大量 cpu 时间，算是符合预期。

如果要改进整体性能，可以优先匹配非正则规则，这样做不会有性能影响，但是依旧解决不了配置不够灵活的问题。 GFWList 的 PAC 中也有正则匹配，可以参考 https://github.com/JinnLynn/genpac 的性能测试。

Jason4774 commented:
![image](https://cloud.githubusercontent.com/assets/19594338/17960175/11b10dd8-6ad7-11e6-941f-6fdcb34e76d4.png)
https://xn--90agrrk8e.xn--p1ai/
就是这个网页，使用cow时正常的，但是Meow就打不开了，虽然可以通过添加条件或者把这个域名加入到强制代理列表来解决，但是我还是觉得如果meow能够直接检测到这个网页需要代理然后就直接代理了，但是为什么这个网页检测不出来呢？是否是因为这个域名是外文的原因呢？
![image](https://cloud.githubusercontent.com/assets/19594338/17960192/294805fa-6ad7-11e6-9b55-d34b6526e0a1.png)
有谁知道这个该怎么解决吗？谢谢
因为感觉cow会影响网速，所以不想用cow，meow速度很快。
